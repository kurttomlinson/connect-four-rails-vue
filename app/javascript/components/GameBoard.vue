<template>
  <div class="board-container">
    <div v-for="(column, columnIndex) in boardState" :key="columnIndex">
        <span v-for="(element, rowIndex) in column" :key="rowIndex" class="board-element">
            {{ element }}
        </span>
    </div>
  </div>
</template>

<script>
import axios from 'axios';

export default {
  props: {
    moves: {
      type: Array,
      required: true,
      validator: function (moves) {
        return moves.reduce(function(valid, move) {
          return valid && !isNaN(move) && (move >= 0) && (move <= 6)
        }, true)
      }
    }
  },
  computed: {
    boardState: function () {
      let columns = []
      let currentPlayer = '1'
      for (let i = 0; i < this.moves.length; i++) {
        let move = this.moves[i]
        if (typeof columns[move]  == 'undefined') {
          columns[move] = []
        }
        columns[move].push(currentPlayer)
        currentPlayer = (currentPlayer == '1') ? '2' : '1'
      }
      for (let i = 0; i < 7; i++) {
        if (typeof columns[i]  == 'undefined') {
          columns[i] = []
        }
        while (columns[i].length < 6) {
            columns[i].push('.')
        }
      }
      // transpose and reverse the array to make it easy to print to screen
      let boardState = columns[0].map((col, i) => columns.map(row => row[i])).reverse();
      return boardState
    }
  }
}
</script>

<style scoped>
.board-container {
    text-align: center;
}
.board-element {
    display: inline-block;
    width: 30px;
    height: 30px;
}
</style>
