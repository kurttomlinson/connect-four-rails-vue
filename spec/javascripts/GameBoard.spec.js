import { shallow } from 'vue-test-utils'
import Vue from 'vue'
import GameBoard from 'components/GameBoard'

describe('GameBoard.vue', () => {
  let wrapper

  beforeEach(() => {
    wrapper = shallow(GameBoard, {
      propsData: {
        moves: []
      }
    })
  })

  it('computes the board state correctly', () => {
    wrapper.setProps({
      moves: [0,1,0,1,3,3,3,3]
    })
    expect(wrapper.vm.boardState).toEqual([[".", ".", ".", ".", ".", ".", "."],
                                           [".", ".", ".", ".", ".", ".", "."],
                                           [".", ".", ".", "2", ".", ".", "."],
                                           [".", ".", ".", "1", ".", ".", "."],
                                           ["1", "2", ".", "2", ".", ".", "."],
                                           ["1", "2", ".", "1", ".", ".", "."]])
  })

  it('matches the snapshot', () => {
    wrapper.setProps({
      moves: [0,1,0,1,3,3,3,3]
    })
    expect(wrapper.element).toMatchSnapshot()
  })
})
