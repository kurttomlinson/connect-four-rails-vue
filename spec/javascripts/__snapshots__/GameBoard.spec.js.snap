// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard.vue matches the snapshot 1`] = `
<div
  class="board-container"
>
  <div>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
  </div>
  <div>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
  </div>
  <div>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          2
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
  </div>
  <div>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          1
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
  </div>
  <div>
    <span
      class="board-element"
    >
      
          1
      
    </span>
    <span
      class="board-element"
    >
      
          2
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          2
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
  </div>
  <div>
    <span
      class="board-element"
    >
      
          1
      
    </span>
    <span
      class="board-element"
    >
      
          2
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          1
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
    <span
      class="board-element"
    >
      
          .
      
    </span>
  </div>
</div>
`;
